core_new/src/message/context.rs

```diff
@@ -0,0 +1,313 @@
+// This file is part of Gear.
+
+// Copyright (C) 2022 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::identifiers::{MessageId, ProgramId};
+use crate::message::{
+    Dispatch, HandleMessage, HandlePacket, IncomingMessage, InitMessage, InitPacket, Payload,
+    ReplyMessage, ReplyPacket,
+};
+use alloc::{
+    collections::{BTreeMap, BTreeSet},
+    vec::Vec,
+};
+use codec::{Decode, Encode};
+use scale_info::TypeInfo;
+
+/// Context settings.
+#[derive(Copy, Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Decode, Encode, TypeInfo)]
+pub struct ContextSettings {
+    sending_fee: u64,
+    outgoing_limit: u32,
+}
+
+impl ContextSettings {
+    /// Create new ContextSettings.
+    pub fn new(sending_fee: u64, outgoing_limit: u32) -> Self {
+        Self {
+            sending_fee,
+            outgoing_limit,
+        }
+    }
+}
+
+impl Default for ContextSettings {
+    fn default() -> Self {
+        Self::new(0, 1024)
+    }
+}
+
+/// Context outcome.
+///
+/// Contains all sendings and wakes that should be done after execution.
+#[derive(Default, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Decode, Encode, TypeInfo)]
+pub struct ContextOutcome {
+    init: Vec<InitMessage>,
+    handle: Vec<HandleMessage>,
+    reply: Option<ReplyMessage>,
+    awakening: Vec<MessageId>,
+    // Additional information section
+    program_id: ProgramId,
+    source: ProgramId,
+    origin_msg_id: MessageId,
+}
+
+impl ContextOutcome {
+    /// Create new ContextOutcome.
+    fn new(program_id: ProgramId, source: ProgramId, origin_msg_id: MessageId) -> Self {
+        Self {
+            program_id,
+            source,
+            origin_msg_id,
+            ..Default::default()
+        }
+    }
+
+    /// Destructs outcome after execution and returns provided dispatches and awaken message ids.
+    pub fn drain(self) -> (Vec<Dispatch>, Vec<MessageId>) {
```

Comment:
I don't like the name `drain` here and I'd use some kind of `into` here.

Reply:
Any other ideas? Into seems not informational. We had drain before so I decided to keep old thing here :D